src/OpenTelemetry.Collector.Dependencies/Implementation/SqlClientDiagnosticListener.cs

```diff
@@ -0,0 +1,124 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+using System;
+using System.Diagnostics;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Collector.Dependencies.Implementation
+{
+    internal class SqlClientDiagnosticListener : ListenerHandler
+    {
+        private readonly PropertyFetcher commandFetcher = new PropertyFetcher("Command");
+        private readonly PropertyFetcher connectionFetcher = new PropertyFetcher("Connection");
+        private readonly PropertyFetcher dataSourceFetcher = new PropertyFetcher("DataSource");
+        private readonly PropertyFetcher databaseFetcher = new PropertyFetcher("Database");
+        private readonly PropertyFetcher commandTextFetcher = new PropertyFetcher("CommandText");
+        private readonly PropertyFetcher exceptionFetcher = new PropertyFetcher("Exception");
+
+        public SqlClientDiagnosticListener(string sourceName, ITracer tracer)
+            : base(sourceName, tracer)
+        {
+        }
+
+        public void OnCompleted()
+        {
+        }
+
+        public void OnError(Exception error)
+        {
+        }
+
+        public override void OnStartActivity(Activity activity, object payload)
+        {
+        }
+
+        public override void OnCustom(string name, Activity activity, object payload)
+        {
+            switch (name)
+            {
+                case "System.Data.SqlClient.WriteCommandBefore":
+                case "Microsoft.Data.SqlClient.WriteCommandBefore":
+                    {
+                        var command = this.commandFetcher.Fetch(payload);
+
+                        const string EventNameSuffix = ".OnStartActivity";
+                        if (command == null)
+                        {
+                            CollectorEventSource.Log.NullPayload(nameof(SqlClientDiagnosticListener) + EventNameSuffix);
+                            return;
+                        }
+
+                        var connection = this.connectionFetcher.Fetch(command);
+                        var dataSource = this.dataSourceFetcher.Fetch(connection);
+                        var database = this.databaseFetcher.Fetch(connection);
+                        var commandText = this.commandTextFetcher.Fetch(command);
+
+                        this.Tracer.StartActiveSpan(database.ToString(), SpanKind.Client, out var span);
+
+                        span.PutComponentAttribute("SQL Server");
+
+                        span.PutDatabaseTypeAttribute("sql");
+                        span.PutDatabaseInstanceAttribute(database.ToString());
+                        span.PutDatabaseStatementAttribute(commandText.ToString());
+                        span.SetAttribute("peer.address", dataSource.ToString());
```

Comment:
According to the specification, we should add `peer.hostname` and `peer.address` But these two attributes seem similar.

Reply:
what is the type of `dataSource` and why it needs `ToString()` ?

Reply:
DataSource is `string`

Reply:
same here, make type explicit. Don't add attribute if fetcher did not return string.